homeassistant/util/executor.py

```diff
@@ -0,0 +1,73 @@
+"""Executor util helpers."""
+
+from concurrent.futures import ThreadPoolExecutor
+import logging
+import queue
+import sys
+import traceback
+
+from homeassistant.util.thread import async_raise
+
+_LOGGER = logging.getLogger(__name__)
+
+_JOIN_ATTEMPTS = 10
+
+START_LOG_ATTEMPT = 2
+
+
+def _log_thread_running_at_shutdown(name: str, ident: int) -> None:
+    """Log the stack of a thread that was still running at shutdown."""
+
+    frames = sys._current_frames()  # pylint: disable=protected-access
+    stack = frames.get(ident)
+    formatted_stack = traceback.format_stack(stack)
+    _LOGGER.warning(
+        "Thread[%s] is still running at shutdown: %s",
+        name,
+        "".join(formatted_stack).strip(),
+    )
+
+
+class InterruptibleThreadPoolExecutor(ThreadPoolExecutor):
+    """A ThreadPoolExecutor instance that will not deadlock on shutdown."""
+
+    def logged_shutdown(self) -> None:
+        """Shutdown backport from cpython 3.9 with interrupt support added."""
+        with self._shutdown_lock:  # type: ignore[attr-defined]
+            self._shutdown = True
+            # Drain all work items from the queue, and then cancel their
+            # associated futures.
+            while True:
+                try:
+                    work_item = self._work_queue.get_nowait()
+                except queue.Empty:
+                    break
+                if work_item is not None:
+                    work_item.future.cancel()
+            # Send a wake-up to prevent threads calling
+            # _work_queue.get(block=True) from permanently blocking.
+            self._work_queue.put(None)
+
+        remaining_threads = list(self._threads)  # type: ignore[attr-defined]
+
+        for attempt in range(1, _JOIN_ATTEMPTS + 1):
+            joined = []
+
+            for thread in remaining_threads:
+                thread.join(timeout=0.1)
+                ident = thread.ident
+                if not thread.is_alive() or ident is None:
+                    joined.append(thread)
+                    continue
+
+                if attempt >= START_LOG_ATTEMPT:
+                    _log_thread_running_at_shutdown(thread.name, ident)
+
+                async_raise(ident, SystemExit)
+                thread.join(timeout=1)
```

Comment:
Since MAX_EXECUTOR_WORKERS is 64, this could potentially wait for ~704s plus execution time if all threads were blocked.  In practice it is usually one or two that are blocked, but on a system that is running some problematic code it is probably a good assumption that a lot more would be.  

https://github.com/home-assistant/supervisor/blob/main/supervisor/docker/homeassistant.py#L44

We should add whatever the maximum time is to the supervisor shutdown timeout to avoid a forced kill.  

Maybe adjust the timeout interval downward based on the number of threads that don't join on the first attempt so we never spend a total of more than 60 seconds in the shutdown?

Reply:
We also need to add the time it could take to shut down non executor threads since there can be any number of those and they could be in a similar state. These are far less likely common since most of the problematic code seems to be run in the executor. This is of course limited to my testing and I don't have a lot of custom components running so I expect it's actually worse in the wild.

Reply:
I'll revise both parts to achieve a maximum timeout and open a PR to supervisor to adjust the timeout with names for each value so we have clarity what the numbers mean

Reply:
I'll work on this after sleep as I want to make sure I'm at my best since getting this right is a bit mission critical

Reply:
Maybe we can have a separate timer for the total time? It sounds good to make sure that all threads get at least one chance to clean up.

Reply:
I've add an overall timer in both places.